src/charm.py

```diff
@@ -35,83 +83,595 @@ def __init__(self, *args: typing.Any):
+        uris = [
+            f"{'https' if uses_tls else 'http'}://{endpoint}"
+            for endpoint in data["endpoints"].split(",")
+        ]
+        env = {
+            "ELASTICSEARCH__URL": json.dumps(uris),
+            "ELASTICSEARCH__INDEX_PREFIX": self.app.name,
+        }
+        if "tls-ca" in data:
+            env["ELASTICSEARCH__SSL__CA"] = "/opt/opencti/config/opensearch.pem"
+        username, password = data.get("username"), data.get("password")
+        if username:
+            env["ELASTICSEARCH__USERNAME"] = username
+        if password:
+            env["ELASTICSEARCH__PASSWORD"] = password
+        return env
+
+    def _extract_opensearch_info(self) -> dict:
+        """Extract opensearch connection information from the opensearch integration.
+
+        Returns:
+            A dictionary containing the opensearch connection info.
+
+        Raises:
+            InvalidIntegration: invalid OpenSearch integration.
+        """
+        integration = typing.cast(
+            ops.Relation, self.model.get_relation(self._opensearch.relation_name)
+        )
+        integration_id = integration.id
+        try:
+            return self._opensearch.fetch_relation_data(
+                relation_ids=[integration_id],
+                fields=["endpoints", "username", "password", "tls", "tls-ca", "index"],
+            )[integration_id]
+        except ops.ModelError as exc:
+            # secret in integration not accessible before the integration events?
+            logger.error(
+                "invalid opensearch-client integration: %s",
+                self._dump_integration("opensearch-client"),
+            )
+            raise InvalidIntegration("invalid opensearch integration") from exc
+
+    def _install_opensearch_cert(self) -> None:
+        """Install opensearch TLS certification obtained from integration to the container.
+
+        Do nothing if opensearch doesn't use TLS.
+        """
+        data = self._extract_opensearch_info()
+        if ca := data.get("tls-ca"):
+            self._container.make_dir("/opt/opencti/config/", make_parents=True)
+            self._container.push(
+                "/opt/opencti/config/opensearch.pem",
+                source=ca,
+                encoding="ascii",
+            )
+
+    def _gen_redis_env(self) -> dict[str, str]:
+        """Generate the Redis-related environment variables for the OpenCTI platform.
+
+        Returns:
+            A dictionary containing the Redis-related environment variables.
+
+        Raises:
+            IntegrationNotReady: redis integration not ready.
+            InvalidIntegration: invalid Redis integration.
+        """
+        redis_url = self._redis.url
+        # bug in the Redis library produces an ill-formed redis_url
+        # when the integration is not ready
+        if not redis_url or redis_url == "redis://None:None":
+            raise IntegrationNotReady("waiting for redis integration")
```

Comment:
Do you think these for all the integrations should be done in the precondition check?

Reply:
The summary of the flow in the charm is: `1. precondition check => 2. data extraction => 3. data validation`. This check belongs to the `3. data validation` step, so it needs to be done after the `2. data extraction`, which follows the `1. precondition check`.